<!--
  2 failure:  :disappointed_rel..., :microscope: Ther...
  1 warning:  There is no linke...
  1 messages
  
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  :microscope: There are library changes, but no changes to the unit tests. That's OK as long as you're refactoring existing code, but will require an admin to merge this PR. Please see <a href="https://github.com/garymathews/titanium_mobile/blob/TIMOB-26202_2/README.md#unit-tests">README.md#unit-tests</a> for docs on unit testing.
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  There is no linked JIRA ticket in the PR body. Please include the URL of the relevant JIRA ticket. If you need to, you may file a ticket on <a href="https://jira.appcelerator.org/secure/CreateIssue!default.jspa">JIRA</a>
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.4.0 test /Users/build/jenkins/workspace/dk_titanium_mobile_PR-10195-KD22U3MS4XMJ6XCHWGUEEYWHVK3WEJ4IILRYKJ2ONMX3ZRDHHZFA
> grunt

Running "appcJs:src:lintOnly" (appcJs) task

Running "eslint:src" (eslint) task

Running "checkFormat:ios" (checkFormat) task

Running "checkFormat:android" (checkFormat) task
Fatal error: Error: Formatting incorrect on "android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='45958' length='4'>&#10;		</replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

--- a/android/kroll-apt/src/java/org/appcelerator/kroll/annotations/generator/ProxyBinding.fm
+++ b/android/kroll-apt/src/java/org/appcelerator/kroll/annotations/generator/ProxyBinding.fm
@@ -720,3 +720,15 @@ beprovided to handle the result.
 		}
 	}
 </#macro>
+
+<#macro getHolder>
+	Local<Object> holder = args.Holder();
+	if (!JavaObject::isJavaObject(holder)) {
+		holder = holder->FindInstanceInPrototypeChain(getProxyTemplate(isolate));
+	}
+	if (holder.IsEmpty() || holder->IsNull()) {
+		LOGE(TAG, "Couldn't obtain argument holder");
+		args.GetReturnValue().Set(v8::Undefined(isolate));
+		return;
+	}
+</#macro>

--- a/android/kroll-apt/src/java/org/appcelerator/kroll/annotations/generator/ProxyBindingV8.cpp.fm
+++ b/android/kroll-apt/src/java/org/appcelerator/kroll/annotations/generator/ProxyBindingV8.cpp.fm
@@ -244,12 +244,18 @@ void ${className}::${method.apiName}(const FunctionCallbackInfo<Value>& args)
 	<@Proxy.initMethodID className=className name=name signature=signature logOnly=false/>
 
 	Local<Object> holder = args.Holder();
-	// If holder isn't the JavaObject wrapper we expect, look up the prototype chain
 	if (!JavaObject::isJavaObject(holder)) {
 		holder = holder->FindInstanceInPrototypeChain(getProxyTemplate(isolate));
 	}
-
+	if (holder.IsEmpty() || holder->IsNull()) {
+		args.GetReturnValue().Set(v8::Undefined(isolate));
+		return;
+	}
 	titanium::Proxy* proxy = NativeObject::Unwrap<titanium::Proxy>(holder);
+	if (!proxy) {
+		args.GetReturnValue().Set(Undefined(isolate));
+		return;
+	}
 
 	<#if method.args?size &gt; 0>
 	<@Proxy.verifyAndConvertArguments method.args method />
@@ -287,8 +293,15 @@ void ${className}::getter_${name}(Local<Name> property, const PropertyCallbackIn
 	<@Proxy.initJNIEnv/>
 	<@Proxy.initMethodID className=className name=property.getMethodName signature=getSignature logOnly=false/>
 
-	titanium::Proxy* proxy = NativeObject::Unwrap<titanium::Proxy>(args.Holder());
-
+	Local<Object> holder = args.Holder();
+	if (!JavaObject::isJavaObject(holder)) {
+		holder = holder->FindInstanceInPrototypeChain(getProxyTemplate(isolate));
+	}
+	if (holder.IsEmpty() || holder->IsNull()) {
+		args.GetReturnValue().Set(v8::Undefined(isolate));
+		return;
+	}
+	titanium::Proxy* proxy = NativeObject::Unwrap<titanium::Proxy>(holder);
 	if (!proxy) {
 		args.GetReturnValue().Set(Undefined(isolate));
 		return;
@@ -324,7 +337,15 @@ void ${className}::setter_${name}(Local<Name> property, Local<Value> value, cons
 
 	<@Proxy.initMethodID className=className name=property.setMethodName signature=setSignature logOnly=true />
 
-	titanium::Proxy* proxy = NativeObject::Unwrap<titanium::Proxy>(args.Holder());
+	Local<Object> holder = args.Holder();
+	if (!JavaObject::isJavaObject(holder)) {
+		holder = holder->FindInstanceInPrototypeChain(getProxyTemplate(isolate));
+	}
+	if (holder.IsEmpty() || holder->IsNull()) {
+		args.GetReturnValue().Set(v8::Undefined(isolate));
+		return;
+	}
+	titanium::Proxy* proxy = NativeObject::Unwrap<titanium::Proxy>(holder);
 	if (!proxy) {
 		return;
 	}
@@ -363,8 +384,15 @@ void ${className}::interceptor(Local<String> property, const v8::PropertyCallbac
 	<@Proxy.initJNIEnv/>
 	<@Proxy.initMethodID className=className name=interceptor.name signature="(Ljava/lang/String;)Ljava/lang/Object;" logOnly=false/>
 
-	titanium::Proxy* proxy = NativeObject::Unwrap<titanium::Proxy>(args.Holder());
-
+	Local<Object> holder = args.Holder();
+	if (!JavaObject::isJavaObject(holder)) {
+		holder = holder->FindInstanceInPrototypeChain(getProxyTemplate(isolate));
+	}
+	if (holder.IsEmpty() || holder->IsNull()) {
+		args.GetReturnValue().Set(v8::Undefined(isolate));
+		return;
+	}
+	titanium::Proxy* proxy = NativeObject::Unwrap<titanium::Proxy>(holder);
 	if (!proxy) {
 		args.GetReturnValue().Set(Undefined(isolate));
 		return;

--- a/android/kroll-apt/src/java/org/appcelerator/kroll/annotations/generator/ProxyBindingV8.cpp.fm
+++ b/android/kroll-apt/src/java/org/appcelerator/kroll/annotations/generator/ProxyBindingV8.cpp.fm
@@ -243,14 +243,7 @@ void ${className}::${method.apiName}(const FunctionCallbackInfo<Value>& args)
 	<@Proxy.initJNIEnv/>
 	<@Proxy.initMethodID className=className name=name signature=signature logOnly=false/>
 
-	Local<Object> holder = args.Holder();
-	if (!JavaObject::isJavaObject(holder)) {
-		holder = holder->FindInstanceInPrototypeChain(getProxyTemplate(isolate));
-	}
-	if (holder.IsEmpty() || holder->IsNull()) {
-		args.GetReturnValue().Set(v8::Undefined(isolate));
-		return;
-	}
+	<@Proxy.getHolder/>
 	titanium::Proxy* proxy = NativeObject::Unwrap<titanium::Proxy>(holder);
 	if (!proxy) {
 		args.GetReturnValue().Set(Undefined(isolate));
@@ -293,14 +286,7 @@ void ${className}::getter_${name}(Local<Name> property, const PropertyCallbackIn
 	<@Proxy.initJNIEnv/>
 	<@Proxy.initMethodID className=className name=property.getMethodName signature=getSignature logOnly=false/>
 
-	Local<Object> holder = args.Holder();
-	if (!JavaObject::isJavaObject(holder)) {
-		holder = holder->FindInstanceInPrototypeChain(getProxyTemplate(isolate));
-	}
-	if (holder.IsEmpty() || holder->IsNull()) {
-		args.GetReturnValue().Set(v8::Undefined(isolate));
-		return;
-	}
+	<@Proxy.getHolder/>
 	titanium::Proxy* proxy = NativeObject::Unwrap<titanium::Proxy>(holder);
 	if (!proxy) {
 		args.GetReturnValue().Set(Undefined(isolate));
@@ -337,14 +323,7 @@ void ${className}::setter_${name}(Local<Name> property, Local<Value> value, cons
 
 	<@Proxy.initMethodID className=className name=property.setMethodName signature=setSignature logOnly=true />
 
-	Local<Object> holder = args.Holder();
-	if (!JavaObject::isJavaObject(holder)) {
-		holder = holder->FindInstanceInPrototypeChain(getProxyTemplate(isolate));
-	}
-	if (holder.IsEmpty() || holder->IsNull()) {
-		args.GetReturnValue().Set(v8::Undefined(isolate));
-		return;
-	}
+	<@Proxy.getHolder/>
 	titanium::Proxy* proxy = NativeObject::Unwrap<titanium::Proxy>(holder);
 	if (!proxy) {
 		return;
@@ -384,14 +363,7 @@ void ${className}::interceptor(Local<String> property, const v8::PropertyCallbac
 	<@Proxy.initJNIEnv/>
 	<@Proxy.initMethodID className=className name=interceptor.name signature="(Ljava/lang/String;)Ljava/lang/Object;" logOnly=false/>
 
-	Local<Object> holder = args.Holder();
-	if (!JavaObject::isJavaObject(holder)) {
-		holder = holder->FindInstanceInPrototypeChain(getProxyTemplate(isolate));
-	}
-	if (holder.IsEmpty() || holder->IsNull()) {
-		args.GetReturnValue().Set(v8::Undefined(isolate));
-		return;
-	}
+	<@Proxy.getHolder/>
 	titanium::Proxy* proxy = NativeObject::Unwrap<titanium::Proxy>(holder);
 	if (!proxy) {
 		args.GetReturnValue().Set(Undefined(isolate));

--- a/android/modules/ui/src/java/ti/modules/titanium/ui/TableViewProxy.java
+++ b/android/modules/ui/src/java/ti/modules/titanium/ui/TableViewProxy.java
@@ -675,6 +675,7 @@ public class TableViewProxy extends TiViewProxy
 	{
 		ArrayList<TableViewSectionProxy> sections = getSectionsArray();
 		for (TableViewSectionProxy section : sections) {
+			section.releaseViews();
 			section.setParent(null);
 		}
 		sections.clear();

--- a/android/modules/ui/src/java/ti/modules/titanium/ui/TableViewRowProxy.java
+++ b/android/modules/ui/src/java/ti/modules/titanium/ui/TableViewRowProxy.java
@@ -272,30 +272,33 @@ public class TableViewRowProxy extends TiViewProxy
 	@Override
 	public void releaseViews()
 	{
-		super.releaseViews();
-
-		if (controls != null) {
-			for (TiViewProxy control : controls) {
-				control.releaseKroll();
-			}
-		}
-	}
-
-	@Override
-	public void release()
-	{
-		super.release();
-
-		releaseViews();
-
 		if (tableViewItem != null) {
 			tableViewItem.release();
 			tableViewItem = null;
 		}
 		if (controls != null) {
+			for (TiViewProxy control : controls) {
+				control.releaseViews();
+			}
+		}
+
+		super.releaseViews();
+	}
+
+	@Override
+	public void release()
+	{
+		releaseViews();
+
+		if (controls != null) {
+			for (TiViewProxy control : controls) {
+				control.release();
+			}
 			controls.clear();
 			controls = null;
 		}
+
+		super.release();
 	}
 
 	public TiTableViewRowProxyItem getTableViewRowProxyItem()

--- a/android/modules/ui/src/java/ti/modules/titanium/ui/android/AndroidModule.java
+++ b/android/modules/ui/src/java/ti/modules/titanium/ui/android/AndroidModule.java
@@ -251,7 +251,7 @@ public class AndroidModule extends KrollModule
 				Activity currentActivity = TiApplication.getAppCurrentActivity();
 				if (currentActivity != null) {
 					TiUIHelper.showSoftKeyboard(currentActivity.getWindow().getDecorView(), false);
-				} else if (activity != null) {
+				} else if (getActivity() != null) {
 					TiUIHelper.showSoftKeyboard(getActivity().getWindow().getDecorView(), false);
 				} else {
 					Log.w(TAG, "Unable to hide soft keyboard. Activity is null", Log.DEBUG_MODE);

--- a/android/modules/ui/src/java/ti/modules/titanium/ui/widget/searchview/TiUISearchView.java
+++ b/android/modules/ui/src/java/ti/modules/titanium/ui/widget/searchview/TiUISearchView.java
@@ -162,4 +162,19 @@ public class TiUISearchView extends TiUIView implements SearchView.OnQueryTextLi
 	{
 		searchChangeListener = listener;
 	}
+
+	@Override
+	public void release()
+	{
+		if (searchView != null) {
+			searchView.setOnQueryTextListener(null);
+			searchView.setOnCloseListener(null);
+			searchView.setOnQueryTextFocusChangeListener(null);
+			searchView = null;
+		}
+		if (searchChangeListener != null) {
+			searchChangeListener = null;
+		}
+		super.release();
+	}
 }

--- a/android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableViewRowProxyItem.java
+++ b/android/modules/ui/src/java/ti/modules/titanium/ui/widget/tableview/TiTableViewRowProxyItem.java
@@ -434,6 +434,11 @@ public class TiTableViewRowProxyItem extends TiBaseTableViewItem
 			}
 		}
 
+		if (content == null) {
+			this.content = new TiCompositeLayout(getRowProxy().getActivity());
+			addView(content, new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT));
+		}
+
 		if (props.containsKey(TiC.PROPERTY_LAYOUT)) {
 			content.setLayoutArrangement(TiConvert.toString(props, TiC.PROPERTY_LAYOUT));
 		}
@@ -647,13 +652,15 @@ public class TiTableViewRowProxyItem extends TiBaseTableViewItem
 	@Override
 	public void release()
 	{
-		super.release();
 		if (views != null) {
 			for (TiUIView view : views) {
 				view.release();
 			}
 			views.clear();
 			views = null;
+		}
+		if (item != null) {
+			item = null;
 		}
 		if (content != null) {
 			content.removeAllViews();
@@ -667,5 +674,7 @@ public class TiTableViewRowProxyItem extends TiBaseTableViewItem
 			hasChildDrawable.setCallback(null);
 			hasChildDrawable = null;
 		}
+
+		super.release();
 	}
 }

--- a/android/runtime/v8/src/java/org/appcelerator/kroll/runtime/v8/ReferenceTable.java
+++ b/android/runtime/v8/src/java/org/appcelerator/kroll/runtime/v8/ReferenceTable.java
@@ -61,10 +61,7 @@ public final class ReferenceTable
 	public static void destroyReference(long key)
 	{
 		Log.d(TAG, "Destroying reference under key: " + key, Log.DEBUG_MODE);
-		Object obj = references.remove(key);
-		if (obj instanceof Reference) {
-			obj = ((Reference<?>) obj).get();
-		}
+		Object obj = getReference(key);
 		// If it's an V8Object, set the ptr to 0, because the proxy is dead on C++ side
 		// This *should* prevent the native code from trying to reconstruct the proxy for any reason
 		if (obj instanceof KrollProxySupport) {
@@ -75,6 +72,10 @@ public final class ReferenceTable
 				v8.setPointer(0);
 			}
 		}
+		if (obj instanceof KrollProxy) {
+			((KrollProxy) obj).setReferenceKey(0);
+		}
+		references.remove(key);
 	}
 
 	/**
@@ -85,7 +86,7 @@ public final class ReferenceTable
 	public static void makeWeakReference(long key)
 	{
 		Log.d(TAG, "Downgrading to weak reference for key: " + key, Log.DEBUG_MODE);
-		Object ref = references.get(key);
+		Object ref = getReference(key);
 		references.remove(key);
 		references.put(key, new WeakReference<Object>(ref));
 	}
@@ -98,7 +99,7 @@ public final class ReferenceTable
 	public static void makeSoftReference(long key)
 	{
 		Log.d(TAG, "Downgrading to soft reference for key: " + key, Log.DEBUG_MODE);
-		Object ref = references.get(key);
+		Object ref = getReference(key);
 		references.remove(key);
 		references.put(key, new SoftReference<Object>(ref));
 	}
@@ -144,6 +145,6 @@ public final class ReferenceTable
 	public static boolean isStrongReference(long key)
 	{
 		Object ref = getReference(key);
-		return !(ref instanceof WeakReference || ref instanceof SoftReference);
+		return !(ref instanceof Reference);
 	}
 }

--- a/android/runtime/v8/src/native/JavaObject.cpp
+++ b/android/runtime/v8/src/native/JavaObject.cpp
@@ -80,7 +80,9 @@ jobject JavaObject::getJavaObject()
 		jobject ref = ReferenceTable::getReference(refTableKey_);
 		if (ref == NULL) {
 			// Sanity check. Did we get into a state where it was weak on Java, got GC'd but the C++ proxy didn't get deleted yet?
-			LOGE(TAG, "!!! OH NO! We tried to grab a Java Object back out of the reference table, but it must have been GC'd, because it's null! Key: %d", refTableKey_);
+			LOGW(TAG, "Could not obtain reference, java object has already been collected! (Key: %d)", refTableKey_);
+			refTableKey_ = 0;
+			javaObject_ = NULL;
 		}
 		return ref;
 	}
@@ -124,7 +126,7 @@ void JavaObject::MakeJSWeak()
 	// but this time, we say call us back as a finalizer so we can resurrect the
 	// object (save it from really being GCd by V8) and move it's Java object twin
 	// to a weak reference in the JVM. (where we can track when that gets GC'd by the JVM to call back and kill this)
-	if (!isDetached()) {
+	if (!isDetached() && !persistent().IsEmpty()) {
 		persistent().SetWeak(this, DetachCallback, v8::WeakCallbackType::kFinalizer); // MUST BE kFinalizer or our object cannot be resurrected!
 		persistent().MarkIndependent();
 	}
@@ -181,7 +183,9 @@ void JavaObject::MakeJavaStrong()
 			jobject stored = ReferenceTable::clearReference(refTableKey_);
 			if (stored == NULL) {
 				// Sanity check. Did we get into a state where it was weak on Java, got GC'd but the C++ proxy didn't get deleted yet?
-				LOGE(TAG, "!!! OH NO! We tried to move a weak Java object back to strong, but it's aleady been GC'd by JVM! We're in a bad state! Key: %d", refTableKey_);
+				LOGW(TAG, "Could not move weak reference to strong, java object has already been collected! (Key: %d)", refTableKey_);
+				refTableKey_ = 0;
+				javaObject_ = NULL;
 			} else {
 				env->DeleteLocalRef(stored);
 			}
@@ -238,15 +242,18 @@ void JavaObject::DeleteJavaRef()
 		} else {
 			env->DeleteGlobalRef(javaObject_);
 		}
-		javaObject_ = NULL;
 	} else {
 		LOGD(TAG, "Deleting ref in ReferenceTable for key: %d, pointer: %p", refTableKey_, this);
 		ReferenceTable::destroyReference(refTableKey_); // Kill the Java side
 		refTableKey_ = 0; // throw away the key
 	}
+	javaObject_ = NULL;
 	// When we're done we should be wrapping nothing!
 	ASSERT(javaObject_ == NULL);
 	ASSERT(refTableKey_ == 0);
+
+	// our java object can be collected, make our JS reference weak too
+	// MakeJSWeak();
 }
 
 }

--- a/android/runtime/v8/src/native/Proxy.cpp
+++ b/android/runtime/v8/src/native/Proxy.cpp
@@ -154,12 +154,13 @@ static void onPropertyChangedForProxy(Isolate* isolate, Local<String> property, 
 	jobject javaValue = TypeConverter::jsValueToJavaObject(isolate, env, value, &javaValueIsNew);
 
 	jobject javaProxy = proxy->getJavaObject();
-	env->CallVoidMethod(javaProxy,
-		JNIUtil::krollProxyOnPropertyChangedMethod,
-		javaProperty,
-		javaValue);
-
-	proxy->unreferenceJavaObject(javaProxy);
+	if (javaProxy != NULL) {
+		env->CallVoidMethod(javaProxy,
+			JNIUtil::krollProxyOnPropertyChangedMethod,
+			javaProperty,
+			javaValue);
+		proxy->unreferenceJavaObject(javaProxy);
+	}
 
 	env->DeleteLocalRef(javaProperty);
 	if (javaValueIsNew) {

--- a/android/runtime/v8/src/native/V8Object.cpp
+++ b/android/runtime/v8/src/native/V8Object.cpp
@@ -98,8 +98,13 @@ Java_org_appcelerator_kroll_runtime_v8_V8Object_nativeFireEvent
 		emitter = TypeConverter::javaObjectToJsValue(V8Runtime::v8_isolate, env, jEmitter).As<Object>();
 	}
 
-	Local<Value> fireEventValue = emitter->Get(EventEmitter::emitSymbol.Get(V8Runtime::v8_isolate));
-	if (!fireEventValue->IsFunction()) {
+	Local<String> symbol = EventEmitter::emitSymbol.Get(V8Runtime::v8_isolate);
+	if (emitter.IsEmpty() || symbol.IsEmpty()) {
+		return JNI_FALSE;
+	}
+
+	Local<Value> fireEventValue = emitter->Get(symbol);
+	if (fireEventValue.IsEmpty() || !fireEventValue->IsFunction()) {
 		return JNI_FALSE;
 	}
 

--- a/android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java
+++ b/android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java
@@ -1571,15 +1571,20 @@ public abstract class TiBaseActivity extends AppCompatActivity implements TiActi
 
 		//LW windows
 		if (window == null && view != null) {
-			view.releaseViews();
 			view.release();
+			view = null;
+		}
+		if (view != null) {
+			view.releaseViews();
 			view = null;
 		}
 
 		if (window != null) {
+			if (windowStack.contains(window)) {
+				removeWindowFromStack(window);
+			}
 			window.closeFromActivity(isFinishing);
 			window.releaseViews();
-			window.releaseKroll();
 			window = null;
 		}
 

--- a/android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java
+++ b/android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java
@@ -617,21 +617,38 @@ public abstract class TiViewProxy extends KrollProxy implements Handler.Callback
 
 	public void releaseViews()
 	{
-		if (view != null) {
-			if (children != null) {
-				for (TiViewProxy p : children) {
-					p.releaseViews();
-				}
+		if (children != null) {
+			for (TiViewProxy p : children) {
+				p.releaseViews();
 			}
+		}
+		if (view != null) {
 			view.release();
 			view = null;
 		}
-		releaseKroll();
 		if (runtimeHandler != null) {
 			runtimeHandler = null;
 		}
+		if (mainHandler != null) {
+			mainHandler = null;
+		}
 		setModelListener(null);
+
+		releaseKroll();
 		KrollRuntime.suggestGC();
+	}
+
+	@Override
+	public void release()
+	{
+		releaseViews();
+
+		if (children != null) {
+			children.clear();
+			children = null;
+		}
+
+		super.release();
 	}
 
 	/**

